### DISCUSSION TITLE
[IntelliJ - MacOS] shortcut "option + [" not working on international keyboard

### DISCUSSION THREAD

**Discussion by giansegato** [community/community#7020] [Copilot Conversations]
On Mac OS, the `option + character` shortcut is used to write special characters (eg. `è`, or `ü`) - more context [here](https://www.webnots.com/option-or-alt-key-shortcuts-to-insert-symbols-in-mac-os-x/).

On IntelliJ, Copilot [rotates](https://github.com/github/copilot-docs/blob/main/docs/jetbrains/gettingstarted.md#getting-started-with-github-copilot-in-jetbrains) between suggestions with `option + [` and `option + ]`. This is impossible, as they both yield `‘`, nullifying the suggestion itself. 

Would it be possible to customize the shortcut?

**Comment by jansorg** [comment]
This should already be possible. Please try to set your own shortcuts at `Preferences > Keymap > Plugins > GitHub Copilot`.
Do you have a suggestion for a better shortcut on macOS?

**Reply by ryanzzff** [reply]
How about `cmd + <` and `cmd + >` ?
What I actually mean is: `cmd + shift + ,` and `cmd + shift + .`

**Reply by FunnyHector** [reply]
But customised keymaps don't seem to work? 

Using latest Macos, M1 chip, latest Rubymine, I changed the shortcut to `cmd + ←` and `cmd + →`, and there is no conflicts, but they don't work. Are there any system shortcut conflicts?

EDIT: Changed to other shortcuts and didn't work either. Is this a working feature?

**Reply by MahatirSaid** [reply]
> But customised keymaps don't seem to work?
> 
> Using latest Macos, M1 chip, latest Rubymine, I changed the shortcut to `cmd + ←` and `cmd + →`, and there is no conflicts, but they don't work. Are there any system shortcut conflicts?
> 
> EDIT: Changed to other shortcuts and didn't work either. Is this a working feature?

I HAVE THE EXACT SAME issue.
If I add a custom shortcut for the next completion, it just doesn't work. 
For example, if I chose Option + A, it will just write å instead.

**Reply by giansegato** [reply]
@jansorg I confirm, it doesn't seem to rotate suggestions. Could it be just because the suggestion is a single one, so there's no apparent rotation?
@ryanzzff I like `,` and `.` too, but there seems to be a conflict (Fold code block) - you can override it though

**Reply by jnsvd** [reply]
Same issue here, a customised keymap seems to have no effect (tried multiple non-conflicting combinations).

**Reply by jansorg** [reply]
@giansegato Yes, it's possible that there are no alternative suggestions. I tested custom shortcuts and this worked well for me. Please note, that alternative suggestions are currently only suported for Java and Python files and not every line of code has alternative suggestions. A later update will support more file types. Possible improvements to notify the status is tracked internally.

**Reply by FunnyHector** [reply]
I don't think that's the case. If I press alt + enter to pop up all suggestions, there are more available. Custom shortcuts definitely not working. Please investigate.

**Reply by jnsvd** [reply]
@jansorg If I press ⌥ + ↩, I'm getting suggestions, so that should not be the problem. Note that I'm testing in Webstorm on `*.java` files.

**Reply by jnsvd** [reply]
@jansorg I've performed some additional testing and in IntelliJ I can use the default keys to get the next suggestion. However, as soon you went through all of them, a ‘ character will be typed. The suggestion then disappears and it's no longer possible to go back to previous suggestions (using ⌥ + [). Going back is possible as long as you still see a suggestion and you requested the next suggestion at least once. If you try to go back on the first suggestion, a “ character will be typed instead.

It seems suggestion navigation is disabled in Webstorm (even for Java files) and there are some minor bugs in the current IntelliJ implementation.

**Reply by giansegato** [reply]
I'm testing on a `*.py` file. 
It's weird because for me, Copilot Solutions just says "No completitions found", but I still see at least one after the cursor (see screenshot below). Maybe this can help debug the problem? Happy to make more tests if needed.

![Screen Shot 2021-11-08 at 14 30 21](https://user-images.githubusercontent.com/5476684/140750578-2aff385f-7a98-4420-a693-bec93489b9ac.png)

**Reply by ant-magic42** [reply]
I have the same issue.  Steps to replicate:
- Use a mac with an "international" keyboard eg. UK
- Use PHPStorm and start typing a function somewhere sensible
- Open copilot window with ⌥ + ↩ to verify there are multiple Copilot suggestions
- option + `[` doesn't work as explained above
- Change to custom shortcuts using `Preferences > Keymap > Plugins > GitHub Copilot`
- Custom shortcuts don't work either

**Reply by LeadroyaL** [reply]
Default,  "option key" + '[' or ']' will print something like this `“‘“‘“‘`.
If I change the keymap, no matter how I change the hotkey, it always doesn't work.

Pycharm 2021.3
MacOS 12.0.1

**Reply by SmelayaPanda** [reply]
For IntelliJ products, probably a more complex keyboard shortcut might be an option:
_Control + Option + ]
Control + Option + [_

three keys are of course more difficult, but in a situation where most of the convenient combinations are already used, we have to choose 🙂

**Reply by felixranesberger** [reply]
I´ve changed my shortcut to 
```
Control + Option + ]
Control + Option + [
```

which works fine 👍🏻

**Reply by jeff1985** [reply]
I used 

    Alt + PageDown
    Alt + PageUp

which also works fine for me.

**Comment by douira** [comment]
I have the same issue on VSCode and there is no option to re-bind the default Option + [ which makes picking an alternative with the keyboard shortcuts impossible. There is no [-key on a German keyboard and pressing Shift + Option + 5 doesn't do anything.

**Reply by manuelwaltschek** [reply]
I bound it to `Shift + Option + 5 ` now since - as you said - these keys are free :)
